main/src/main/scala/sbt/Defaults.scala

```diff
@@ -855,9 +855,12 @@ object Defaults extends BuildCommon {
       }
       val trl = (testResultLogger in (Test, test)).value
       val taskName = Project.showContextKey(state.value).show(resolvedScoped.value)
+      val currentLoader = Thread.currentThread.getContextClassLoader
       try {
+        Thread.currentThread.setContextClassLoader(testLoader.value)
```

Comment:
Are we effectively using context classloader as a thread-local global variable here?

Reply:
In other words, is there anything magical about `Thread.currentThread.getContextClassLoader` or could you potentially use any thread-safe global var to pass in `testLoader.value` into `loadClass(...)`?

Reply:
There is something magical about `Thread.getContextClassLoader`. If new threads are spawned by the thread, they inherit the context class loader of the parent. This means that if the test spawns background threads to do work, then the same context class loader is available on all of the background threads. When we set the context class loader to the test loader and then invoke the test runner, the tests run in an environment that is sandbox-like. It's analogous to a forked jvm process with the test loader as the app classloader (though I'm not sure how exactly this analogy holds).

If we used a thread local variable, it wouldn't get copied to threads spawned by the current thread so class loading would likely break somewhere.

FWIW, we already swapped out the context class loader in ScalaRun. I'm not sure why it was done there but not for tests.

Reply:
> If new threads are spawned by the thread, they inherit the context class loader of the parent. 

cool. Thanks for the explanation. Maybe we can add the above as comment.

Reply:
I added a length comment in https://github.com/sbt/sbt/pull/4769/commits/625470cdd5e219ff0e1436c6d157c03e41c935cb